### PR TITLE
planner, unistore: support index scan for common handle on unique index

### LINKED ISSUE #18284: clustered index: 'invalid codec version' when select with unique key
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
set @@tidb_enable_clustered_index=true;
drop table if exists t;
create table t (a varchar(64), b varchar(64), uk int, v int, primary key(a, b), unique key uuk(uk));
insert t values ('a', 'a1', 1, 11), ('b', 'b1', 2, 22), ('c', 'c1', 3, 33);
select * from t use index (uuk);
```

### 2. What did you expect to see? (Required)
```
+---+----+----+----+
| a | b  | uk | v  |
+---+----+----+----+
| a | a1 | 1  | 11 |
| b | b1 | 2  | 22 |
| c | c1 | 3  | 33 |
+---+----+----+----+
```

### 3. What did you see instead (Required)

```console
(1105, 'invalid codec version')
```

### 4. Affected version (Required)

```
Release Version: v4.0.0-beta.2-690-g3f6f36f74-dirty
Edition: Community
Git Commit Hash: 3f6f36f742a2f971cc82feccd04bceb5abc822b3
Git Branch: HEAD
UTC Build Time: 2020-06-29 02:06:58
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #18284 

Problem Summary:

There are 6 index value layouts in total:
  - (> 9 bytes) new collation non-unique
  - (> 9 bytes) new collation unique int handle
  - (> 9 bytes) new collation unique common handle
  - (<= 9 bytes) old collation non-unique
  - (<= 9 bytes) old collation unique int handle
  - (> 9 bytes) old collation unique common handle

In unistore, the value length is not enough to distinguish whether an index value is a new collation or not. Thus, __Old collation unique common handle__ is processed incorrectly.

https://github.com/pingcap/tidb/blob/ffc1bcaaa7fc49e8ef173cd8d2dd0c2a1a5ebc4a/store/mockstore/unistore/cophandler/closure_exec.go#L625-L630

This PR adds common handle support for `indexScanProcessNewCollation`.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
- reuse the logic of `DecodeIndexKV` in `tablecodec`.
- add a simple test case.

How it Works:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- No release note

### CODE REVIEW

File: tablecodec/tablecodec.go [hunk 1]
```diff
@@ -788,12 +785,15 @@ func decodeIndexKVUniqueCommonHandle(key, value []byte, colsLen int, hdStatus Ha
 	if hdStatus == HandleNotNeeded {
 		return resultValues, nil
 	}
-	// reEncode the common handle to a bytes column
-	reEncodedHandle, err1 := codec.EncodeValue(nil, nil, types.NewBytesDatum(value[4:handleEndOff]))
-	if err1 != nil {
-		return nil, err1
+	cHandle, err := decodeHandleInIndexKey(value[4:handleEndOff])
```
Comment (coocood) [inline tablecodec/tablecodec.go:788]: Does the 'c' means `common`?

**Comment by coocood** [comment]
LGTM

File: tablecodec/tablecodec.go [hunk 2]
```diff
@@ -686,15 +692,15 @@ func decodeIndexKvNewCollation(key, value []byte, hdStatus HandleStatus, columns
 		if err != nil {
 			return nil, err
 		}
-		resultValues = append(resultValues, handleBytes)
+		resultValues = append(resultValues, handleBytes...)
 	} else {
-		// In unique index.
+		// In unique int handle index.
```
Comment (wjhuang2016) [inline tablecodec/tablecodec.go:697]: How to decode unique common handle?

File: tablecodec/tablecodec.go [hunk 3]
```diff
@@ -728,28 +734,19 @@ func decodeIndexKvOldCollation(key, value []byte, colsLen int, hdStatus HandleSt
...
-		// unique index
-		handle, err = decodeHandleInIndexValue(value)
+		handleBytes, err := reEncodeHandle(handle, hdStatus == HandleIsUnsigned)
 		if err != nil {
-			return nil, err
+			return nil, errors.Trace(err)
 		}
+		resultValues = append(resultValues, handleBytes...)
+	} else {
+		// In unique int handle index.
```
Comment (wjhuang2016) [inline tablecodec/tablecodec.go:743]: Where is the branch for commom index?

File: tablecodec/tablecodec.go [hunk 4]
```diff
@@ -647,17 +647,23 @@ const (
 )
 
 // reEncodeHandle encodes the handle as a Datum so it can be properly decoded later.
-// If it is common handle, it is encoded as Bytes Datum.
+// If it is common handle, it is encoded as a list of Bytes Datum.
```
Comment (wjhuang2016) [inline tablecodec/tablecodec.go:650]: Why do we need to change this?

Comment (coocood) [inline tablecodec/tablecodec.go:650]:
```suggestion
// If it is common handle, it returns the encoded column values.
```

**Comment by coocood** [comment]
failure caused by env `dial tcp: lookup goproxy.onetool.net on 10.233.0.10:53: no such host`

**Pull request merged**